packages/ui/src/proposals/modals/AddNewProposal/AddNewProposalModal.tsx

```diff
@@ -47,27 +65,66 @@ export const AddNewProposalModal = () => {
     }
 
     if (feeInfo && feeInfo.canAfford) {
-      return setState('WARNING')
+      return send('NEXT')
     }
 
     if (feeInfo && !feeInfo.canAfford) {
-      setState('REQUIREMENTS_FAIL')
+      return send('FAIL')
     }
   }, [state, member?.id, JSON.stringify(feeInfo)])
 
   if (!member || !feeInfo) {
     return null
   }
 
-  if (state === 'REQUIREMENTS_FAIL') {
+  if (state.matches('requirementsFailed')) {
     return (
       <InsufficientFundsModal onClose={hideModal} address={member.controllerAccount} amount={feeInfo.transactionFee} />
     )
   }
 
-  if (state === 'WARNING') {
-    return <AddNewProposalWarningModal onNext={() => setState('PREPARE')} />
+  if (state.matches('warning')) {
+    return <WarningModal onNext={() => send('NEXT')} />
+  }
+
+  if (state.matches('error')) {
+    return <FailureModal onClose={hideModal}>There was a problem while creating proposal.</FailureModal>
+  }
+
+  function selectType(type: ProposalDetails) {
+    setValid(true)
+    setType(type)
```

Comment:
I think that we can persist the `type` in the machine's `context`. Adding typings to the machine will allow one to reliably access it in the later steps.

Reply:
Example on the visualizer: https://xstate.js.org/viz/?gist=559858016b9fb97e46c8982617d267d1

Reply:
✅